functions/index.js

```diff
@@ -91,3 +93,38 @@ exports.TrackUsers = functions.https.onRequest(async (req, res) => {
       
 });
 
+function logUnfetchedCourseCode(courseCode) {
+  console.log("Unable to fetch course data for course code: ", courseCode);
+}
+
+exports.returnText = functions.https.onCall(data => {
+  let courseCodes = data.courseCodes.map(a => a.toUpperCase());
+  let courses = [];
+
+  // Assume rosters is sorted from least recent to most recent semester
+  let rosters = Object.keys(filteredAllCourses);
+  // Iterate over rosters starting from most recent semester
+  for (let i = rosters.length - 1; i >= 0 && courseCodes.length > 0; i -= 1) {
+    let rosterCourses = filteredAllCourses[rosters[i]];
+    // Fetch course objects whose code is in courseCodes
+    let fetchedCourses = rosterCourses.filter(rosterCourse =>
+      courseCodes.indexOf(rosterCourse.subject.concat(' ', rosterCourse.catalogNbr))!= -1);
+    
+    // Delete course codes of fetchedCourses from courseCodes
+    // Update courses with fetchedCourses
+    fetchedCourses.forEach((fetchedCourse) => {
+      let fetchedCourseCode = fetchedCourse.subject.concat(' ', fetchedCourse.catalogNbr);
+      courseCodes = courseCodes.filter(courseCode => courseCode != fetchedCourseCode);
+
+      courses.push(fetchedCourse);
+    });
+
+  }
+  // Log courses that could not be fetched
+  courseCodes.map(a => logUnfetchedCourseCode(a));
+  
+  return {
+    courses: courses
+  };
+});
+
```

Comment:
What's the reason that this is happening on Firebase Functions? An onCall function is just an API, but it also doesn't look like you're using admin privileges to the db. Can't this happen on the frontend?

Reply:
My reasoning was that importing the filtered-all-courses.json would be too heavy a load for the frontend to handle. Also, when a Firebase function is deployed, Firebase seems to package the `functions` folder (that I added filtered-all-courses.json to) to upload once, which I thought would save on time.